test/distributed/fsdp/test_fsdp_summon_full_params.py

```diff
@@ -0,0 +1,264 @@
+from torch.distributed._fsdp import CPUOffload
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import (
+    FSDPTest,
+)
+from torch.testing._internal.common_utils import (
+    TEST_WITH_DEV_DBG_ASAN,
+    run_tests,
+    instantiate_parametrized_tests,
+    parametrize,
+)
+
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class TestSummonFullParams(FSDPTest):
+    @property
+    def world_size(self):
+        return 2
+
+    def get_model_param_count(self, m):
+        return sum([p.numel() for p in m.parameters()])
+
+    # padding ensures that all shards have the same size with the least amount of padding
+    def get_expected_sharded_size(self, global_size):
+        return int(math.ceil(global_size / self.world_size))
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "writeback",
+        [True, False]
+    )
+    @parametrize(
+        "cpu_offload",
+        [CPUOffload(offload_params=True), CPUOffload(offload_params=False)]
+    )
+    @parametrize(
+        "modify_outer",
+        [True, False]
+    )
+    def test_summon_full_param_writeback(self, writeback, cpu_offload, modify_outer):
+        model = FSDP(nn.Sequential(
+            FSDP(nn.Linear(5, 5, bias=False)),
+            nn.Linear(5, 3, bias=False)
+        )).cuda(self.rank)
+
+        # set the value
+        outer_param = model.get_parameter("_fsdp_wrapped_module.flat_param")
+        inner_param = model.get_parameter("_fsdp_wrapped_module._fpw_module.0._fsdp_wrapped_module.flat_param")
+        p = outer_param if modify_outer else inner_param
+
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with model._summon_full_params(writeback=writeback):
+            with torch.no_grad():
+                p.copy_(torch.zeros_like(p))
+
+        if writeback:
+            self.assertEqual(p[0], 0)
+        else:
+            self.assertEqual(p[0], self.rank + 2)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_shard_value(self):
+
+        raw_model = nn.Linear(10, 11)
+        raw_model_size = self.get_model_param_count(raw_model)
+        expected_shard_size = self.get_expected_sharded_size(raw_model_size)
+
+        model = FSDP(raw_model.cuda(self.rank))
+        self.assertEqual(expected_shard_size, self.get_model_param_count(model))
+
+        # we're assuming a single flatenned param
+        self.assertEqual(1, len(list(model.parameters())))
+
+        my_shard = torch.clone(next(model.parameters()))
+
+        with model._summon_full_params():
+            self.assertEqual(raw_model_size, self.get_model_param_count(model))
```

Comment:
should we have some checking such that we ensure the param size is (10, 11) as expected?

Reply:
This is done by how `raw_model_size` is computed by counting the size of a non-FSDP model.